File: JavaScript/19_Binary Tree-III/H124. Binary Tree Maximum Path Sum.js
/**
 * Binary Tree Maximum Path Sum
 * https://leetcode.com/problems/binary-tree-maximum-path-sum/
 *
 * A path in a binary tree is a sequence of nodes where each pair of adjacent nodes
 * in the sequence has an edge connecting them. A node can only appear in the sequence at most once.
 * Note that the path does not need to pass through the root.
 *
 * The path sum of a path is the sum of the node's values in the path.
 *
 * Type             | Details
 * -----------------+----------------------------------------------------
 * Time Complexity  | O(N) – Each node is visited exactly once
 * Space Complexity | O(H) – Recursion stack height (H = height of tree)
 */

/**
 * Definition for a binary tree node.
 */
/**
 * Creates a tree node
 */
function createTreeNode(val = 0, left = null, right = null) {
    return { val, left, right };
}

/**
 * Find the maximum path sum in a binary tree
 * @param {TreeNode} root - Root of the binary tree
 * @returns {number} - Maximum path sum
 */
function maxPathSum(root) {
    let maxSum = -Infinity;

    function dfs(node) {
        if (node === null) return 0;

        const left = Math.max(0, dfs(node.left));
        const right = Math.max(0, dfs(node.right));

        maxSum = Math.max(maxSum, node.val + left + right);

        return node.val + Math.max(left, right);
    }

    dfs(root);
    return maxSum;
}

// Test cases
const root = createTreeNode(1);
root.left = createTreeNode(2);
root.right = createTreeNode(3);

const result = maxPathSum(root);
console.log("Maximum Path Sum:", result); // Output: 6

File: Java/19_Binary Tree-III/H124. Binary Tree Maximum Path Sum.java
// Binary Tree Maximum Path Sum
// https://leetcode.com/problems/binary-tree-maximum-path-sum/description/ 

/**
 * A path in a binary tree is a sequence of nodes where each pair of adjacent nodes
 * in the sequence has an edge connecting them. A node can only appear in the sequence at most once.
 * Note that the path does not need to pass through the root.
 
 * The path sum of a path is the sum of the node's values in the path.
 
 * Given the root of a binary tree, return the maximum path sum of any non-empty path.
 */
 
 class Solution {
     public static class TreeNode {
         int val;
         TreeNode left;
         TreeNode right;
         TreeNode() {}
         TreeNode(int val) { this.val = val; }
         TreeNode(int val, TreeNode left, TreeNode right) {
             this.val = val;
             this.left = left;
             this.right = right;
         }
     }
     
     public int maxPathSum(TreeNode root) {
        int[] maxSum = new int[1];
        maxSum[0] = Integer.MIN_VALUE;
        maxPathSum(root, maxSum);
        return maxSum[0];
     }
     
     private int maxPathSum(TreeNode node, int[] maxSum) {
        if (node == null) return 0;
        
        int left = Math.max(0, maxPathSum(node.left, maxSum));
        int right = Math.max(0, maxPathSum(node.right, maxSum));
        
        maxSum[0] = Math.max(maxSum[0], node.val + left + right);
        
        return node.val + Math.max(left, right);
     }
     
     public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        
        Solution solution = new Solution();
        int maxPathSum = solution.maxPathSum(root);
        System.out.println("Maximum Path Sum: " + maxPathSum);
     }
 }